Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -239,21 +252,279 @@ FieldValue Serializer::DecodeFieldValue(const uint8_t* bytes, size_t length) {
 
   switch (tag) {
     case google_firestore_v1beta1_Value_null_value_tag:
-      DecodeNull(&stream);
+      DecodeNull(stream);
       return FieldValue::NullValue();
     case google_firestore_v1beta1_Value_boolean_value_tag:
-      return FieldValue::BooleanValue(DecodeBool(&stream));
+      return FieldValue::BooleanValue(DecodeBool(stream));
     case google_firestore_v1beta1_Value_integer_value_tag:
-      return FieldValue::IntegerValue(DecodeInteger(&stream));
+      return FieldValue::IntegerValue(DecodeInteger(stream));
     case google_firestore_v1beta1_Value_string_value_tag:
-      return FieldValue::StringValue(DecodeString(&stream));
+      return FieldValue::StringValue(DecodeString(stream));
+    case google_firestore_v1beta1_Value_map_value_tag:
+      return FieldValue::ObjectValue(DecodeObject(stream));
 
     default:
       // TODO(rsgowman): figure out error handling
       abort();
   }
 }
 
+/**
+ * Encodes a FieldValue *and* its length.
+ *
+ * When encoding a top level message, protobuf doesn't include the length (since
+ * you can get that already from the length of the binary output.) But when
+ * encoding a sub/nested message, you must include the length in the
+ * serialization.
+ *
+ * Call this method when encoding a non top level FieldValue. Otherwise call
```

Comment:
This seems kind of oxymoronic because the reality is that outside of tests there never will be a top-level FieldValue. It seems strange to have to write two versions of EncodeFoo and EncodeNestedFoo just to have to make this work.

Meanwhile, it seems like this gunk about

  * create sizing substream
  * call some other code to actually write values
  * take the size
  * write the size
  * if the stream is a sizing substream, inform it of the size
  * otherwise
    * ensure space
    * new substream
    * call user code again to do it for real
    * adjust bytes written and propgate errors

If you took "user code" as a `std::function` that took a `pb_ostream_t*` then this whole function largely becomes an `EncodeNestedMessage(pb_ostream_t* stream, std::function<void(pb_ostream_t*)>)` and no longer specific to FieldValue at all.

Note: happy to address this during the second message type where you find yourself writing this triplet with different types.

Reply:
Yeah, I like that approach. I'll do that as soon as we have a second message like this.